pallets/payroll/src/lib.rs

```diff
@@ -0,0 +1,368 @@
+        CancelStream(
+            StreamId,
+            AccountOf<T>,
+            AccountOf<T>,
+            AssetIdOf<T>,
+            BalanceOf<T>,
+            BalanceOf<T>,
+        ),
+    }
+
+    /// Next Stream Id
+    #[pallet::storage]
+    #[pallet::getter(fn next_stream)]
+    pub type NextStreamId<T: Config> = StorageValue<_, StreamId, ValueQuery>;
+
+    /// Global storage for streams
+    #[pallet::storage]
+    #[pallet::getter(fn get_stream)]
+    pub type Streams<T: Config> = StorageMap<_, Twox64Concat, StreamId, Stream<T>>;
+
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Create a stream
+        #[pallet::weight((<T as Config>::WeightInfo::create_stream(), DispatchClass::Operational))]
+        #[transactional]
+        pub fn create_stream(
+            origin: OriginFor<T>,
+            recipient: AccountOf<T>,
+            deposit: BalanceOf<T>,
+            currency_id: AssetIdOf<T>,
+            rate_per_sec: BalanceOf<T>,
+            start_time: Timestamp,
+            stop_time: Timestamp,
+        ) -> DispatchResultWithPostInfo {
+            let sender = ensure_signed(origin)?;
+            ensure!(recipient != sender, Error::<T>::RecipientIsAlsoSender);
+            ensure!(deposit > 0, Error::<T>::DepositIsZero);
+            ensure!(
+                start_time >= T::UnixTime::now().as_secs(),
+                Error::<T>::StartBeforeBlockTime
+            );
+            ensure!(stop_time > start_time, Error::<T>::StopBeforeStart);
+            // insert stream to the Streams
+            let stream: Stream<T> = Stream {
+                remaining_balance: deposit, // remaining balance same value for now due to initialization
+                deposit,                    // deposit
+                currency_id,                // currency id
+                rate_per_sec,               // rate per  millisecond
+                recipient: recipient.clone(), // recipient
+                sender: sender.clone(),     // sender
+                start_time,                 // start_time
+                stop_time,                  // stop_time
+            };
+            let stream_id = NextStreamId::<T>::get();
+            // Insert stream to runtime
+            Streams::<T>::insert(stream_id, stream);
+            // Increment stream id
+            NextStreamId::<T>::set(stream_id + 1);
+            // transfer deposit from sender to global EOA
+            T::Assets::transfer(currency_id, &sender, &Self::account_id(), deposit, false)?;
+            Self::deposit_event(Event::<T>::CreateStream(
+                stream_id,
+                sender,
+                recipient,
+                deposit,
+                currency_id,
+                start_time,
+                stop_time,
+            ));
+            Ok(().into())
```

Comment:
Would it make sense to return the `stream_id`? 

How would a caller know what the stream ID of their stream is?

Reply:
Usually ppl use indexer(e.g. subquery, subgraph in eth) to know which payroll he has. Therefore, it makes sense to retuen `stream_id` in `create_stream` as it needs to emit an event that triggers indexer to which payroll with the id is assigned to each sender and receiver.

Reply:
A caller(sender) uses an indexer to know what the stream ID they issued just as ethereum uses subgraph, terra uses mantle, and polkadot uses subquery 🤷‍♂️

Reply:
Also, those stored `stream_id` is used to interact with other functions, so it is necessary.